app/src/main/java/com/duckduckgo/app/httpsupgrade/api/HttpsUpgradeDataDownloader.kt

```diff
@@ -0,0 +1,107 @@
+import com.duckduckgo.app.httpsupgrade.db.HttpsWhitelistDao
+import com.duckduckgo.app.httpsupgrade.model.HttpsBloomFilterSpec
+import com.duckduckgo.app.httpsupgrade.model.HttpsBloomFilterSpec.Companion.HTTPS_BINARY_FILE
+import io.reactivex.Completable
+import io.reactivex.Completable.fromAction
+import timber.log.Timber
+import java.io.IOException
+import javax.inject.Inject
+
+class HttpsUpgradeDataDownloader @Inject constructor(
+    private val service: HttpsUpgradeService,
+    private val httpsUpgrader: HttpsUpgrader,
+    private val httpsBloomSpecDao: HttpsBloomFilterSpecDao,
+    private val whitelistDao: HttpsWhitelistDao,
+    private val binaryDataStore: BinaryDataStore
+) {
+
+    fun download(): Completable {
+
+        val filter = service.httpsBloomFilterSpec()
+            .flatMapCompletable {
+                downloadBloomFilter(it)
+            }
+        val whitelist = downloadWhitelist()
+
+        return Completable.mergeDelayError(listOf(filter, whitelist))
+            .doOnComplete {
+                Timber.i("Https download task completed successfully")
+            }
+    }
+
+    private fun downloadBloomFilter(specification: HttpsBloomFilterSpec): Completable {
+        return fromAction {
+            Timber.d("Downloading https bloom filter binary")
+            val call = service.httpsBloomFilter()
+            val response = call.execute()
+            val fileName = HTTPS_BINARY_FILE
+
+            if (response.isCached && binaryDataStore.verifyCheckSum(fileName, specification.sha256)) {
+                Timber.d("Https bloom data already cached and stored for this spec")
+                return@fromAction
+            }
+
+            if (!response.isSuccessful) {
+                throw IOException("Status: ${response.code()} - ${response.errorBody()?.string()}")
+            }
+
+            val bytes = response.body()!!.bytes()
+            if (!binaryDataStore.verifyCheckSum(bytes, specification.sha256)) {
+                throw IOException("Https binary has incorrect checksum, throwisng away file")
+            }
+
+            Timber.d("Updating https bloom data store with new data")
+            httpsBloomSpecDao.insert(specification)
+            binaryDataStore.saveData(fileName, bytes)
+            httpsUpgrader.reloadData()
+        }
+    }
+
+    private fun downloadWhitelist(): Completable {
+
+        Timber.d("Downloading HTTPS whitelist")
+        return fromAction {
+
+            val call = service.whitelist()
+            val response = call.execute()
+
+            if (response.isCached && whitelistDao.count() > 0) {
+                Timber.d("Https whitelist already cached and stored")
+                return@fromAction
+            }
+
+            if (response.isSuccessful) {
+                val whitelist = response.body()!!
+                Timber.d("Updating https whitelist with new data")
+                whitelistDao.deleteAll()
```

Comment:
Should we make a method on the `Dao` which both deletes and inserts, so that we can mark it with `@Transaction` to ensure the deletion and insertion both either happen or neither happens?